### DISCUSSION TITLE
Fixed modes in  non_negative_tucker_hals

### DISCUSSION THREAD

**Discussion by fosfrancesco** [tensorly/tensorly#392] [Q&A] [answered]
Hi tensorly team,

I'm trying to run non negative tucker decomposition (` non_negative_tucker_hals`) with fixed modes. 
My goal would be:
1. run it on some 3-dimensional data D to obtain a decomposition D = [G; A, B, C] 
2. fix A and run it again on some other data to obtain the new matrices G, B, and C (and of course the same A).

If I'm not mistaken, the parameter ` fixed_modes` is there for this reason. So I would set it as `fixed_modes = [2]`.
However, I can't find a way to input the "old" A. By looking at pull requests I think I need to use the `init` parameter, but I can't find the correct way to set it.

Here is how I would like my code to look (there is an "?" that need to be filled with correct values):
```
from tensorly.decomposition import non_negative_tucker_hals

m1 = np.random.random([10,15,20])
tensor_hals, error = non_negative_tucker_hals(a, rank = (3,4,5), return_errors= True)

m2 = np.random.random([10,15,20])
tensor_hals2, error2 = non_negative_tucker_hals(b,rank = (3,4,5), return_errors= True, fixed_modes = [2], init = ?)
```

Thank you for your help and your great work!

Francesco

**Comment by caglayantuna** [comment, marked as answer]
Hi @fosfrancesco ,

Thanks to your question, I realized a bug in `non_negative_tucker_hals` which doesn't work with fixed modes. I will work on it soon and inform you from here.

For the second part of your question, you should use TuckerTensor as init when you use fixed modes or factors (these names should be compatible as well). Here is an example by using your code and `tucker` decomposition:

```python
from tensorly.decomposition import non_negative_tucker_hals, tucker
import numpy as np

m1 = np.random.random([10,15,20])
tensor_hals, error = non_negative_tucker_hals(m1, rank = (3,4,5), return_errors= True)

m2 = np.random.random([10,15,20])
tensor_hals2 = tucker(m2, rank=(3,4,5), fixed_factors= [1], init=tensor_hals)
```
By the way, tucker decomposition has no `return_errors` option that should be included.

Thanks,
Caglayan

**Reply by fosfrancesco** [reply]
Thank you very much for the fast answer, is it clear now!
And let me know when you have time to solve the issue, unfortunately, I really need the non-negativity for my application and I see that `non_negative_tucker` (without hals) does not support the fixed_modes.

Best,

Francesco

**Reply by caglayantuna** [reply]
Hello again,

Actually, `non_negative_tucker_hals` only fails when the last mode is fixed as parafac. So, in your fist question, it was 2 (last mode) and failed. There should be a warning about it or I have another idea to fix the last mode, maybe I can do a PR soon. 

You can run the same example above with `fixed_modes= [1]` but I suggest you to use `init=tensor_hals.tucker_copy()` to avoid changes on `tensor_hals`.

Best,

Caglayan

**Reply by fosfrancesco** [reply]
Interesting! So if I am interested in fixing the third mode of m1, can I just transpose m1 (`np.transpose(m1, (0,2,1))`) and then use fixed_modes = [1]?

Will it be the same thing as fixing [2] on the untransposed matrix? I'm just starting working with those tensor transformations and I think this is correct, but I'm not 100% sure.

**Reply by cohenjer** [reply]
Exactly, you can use transpose beforehand. The last mode thing is a problem that comes from storing the results of the last updated mode to speed up computations of errors and such; we could get rid of it but nobody has the taken the time to do so I guess !

**Reply by fosfrancesco** [reply]
Thank you very much for your help. I will repay you with a lot of paper citations! And some contributions to the repo when I will have the time.
